clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -209,7 +209,10 @@ public RecordAppendResult append(TopicPartition tp,
                     return appendResult;
                 }
 
-                MemoryRecordsBuilder recordsBuilder = recordsBuilder(buffer, maxUsableMagic);
+                // We have to pass `size` because the buffer returned from `free.allocate` could, in theory, be larger
+                // than the requested size if `deallocate(ByteBuffer buffer, int size)` is invoked with a buffer larger
+                // than the `size` parameter
+                MemoryRecordsBuilder recordsBuilder = recordsBuilder(buffer, maxUsableMagic, size);
```

Comment:
Btw, the fact that `size` can be different from `batch.size` means that we won't benefit from `BufferPool` in those cases. But I guess that these cases should be rare.